src/Util/Visitor/VariableCollectionVisitor.php

```diff
@@ -0,0 +1,287 @@
+
+            // parameters count as available variables.
+            $this->variables[] = ReflectionVariable::createFromName($param->name, $type, $param->getAttribute('startLine'));
+        }
+    }
+
+    private function processAssign(Expr\Assign $node)
+    {
+        // assignment is not directly to a variable, so just ignore it
+        // rather than trying to recreate state.
+        if (false === $node->var instanceof Expr\Variable) {
+            return;
+        }
+
+        $type = $this->typeFromExpression($node->expr);
+
+        $this->variables[] = ReflectionVariable::createFromName($node->var->name, $type, $node->getAttribute('startLine'));
+    }
+
+    private function typeFromExpression(Node $expr): ReflectionType
+    {
+        if ($expr instanceof Expr\New_) {
+            return $this->createTypeFromNameNode($expr->class);
+        }
+
+        if ($expr instanceof Expr\PropertyFetch) {
+            $type = $this->typeFromExpression($expr->var);
+
+            if ($type->getTypeObject() instanceof DocType\Object_) {
+                $reflection = $this->reflector->reflect($type);
+                $propertyRefl = $reflection->getProperty($expr->name);
+
+                if ($propertyRefl->getDocComment()) {
+                    $types = $propertyRefl->getDocBlockTypes();
+
+                    return $this->createReflectionTypeFromDocType(reset($types));
+                }
+            }
+
+            return $this->createReflectionTypeFromString('mixed');
+        }
+
+        if ($expr instanceof Expr\MethodCall) {
+            $type = $this->typeFromExpression($expr->var);
+
+            if ($type->getTypeObject() instanceof DocType\Object_) {
+                $reflection = $this->reflector->reflect($type);
+                $method = $reflection->getMethod($expr->name);
+
+                return $method->getReturnType() ?: $this->createReflectionTypeFromString('mixed');
+            }
+
+            return $this->createReflectionTypeFromString('mixed');
+        }
+
+        if ($expr instanceof Expr\Variable) {
+            if ($expr->name === 'this') {
+                $type = $this->createReflectionTypeFromString($this->reflection->getName());
+
+                return $type;
+            }
+
+            if (isset($this->methodParamTypes[$expr->name])) {
+                return $this->methodParamTypes[$expr->name];
+            }
+        }
+
+        // if this is a function call, try and instantiate the runtime native
+        // \ReflectionFunction, if it is not internal then ignore it as we
+        // cannot guarantee that we are running in the same process as the code
+        // we are analyzing.
+        //
+        // TODO: This is no positive test case for this ... which PHP internal functions
+        //       actually have a return type??
+        if ($expr instanceof Expr\FuncCall) {
+            $func = (string) $expr->name;
+            $reflection = new \ReflectionFunction($func);
```

Comment:
You can stop at the return type declared on the function btw - no need to introspect.